Support Python 3.13

eoyoa opened the pull request:
When trying to setup a new dev environment in Python 3.13, you encounter a couple issues:

- `asyncpg@0.28.0` does not work with Python 3.13.
   - `asyncpg@^0.30.0` is compatible: see MagicStack/asyncpg#1188
- `asyncore`, `imghdr`, and `cgi` were all deprecated in 3.12.
   - You can use `python-deadlibs` as a temporary stand-in.

This PR updates these dependencies to their compatible versions to make `rs` compatible with Python 3.13, and also bumps versions of affected projects.

Note that currently, you are unable to test build this project since `rsmanage` requires the newly bumped 7.7.4 version of `runestone`, which is not on PyPi until @bnmnetp updates the project. However, it should work in theory, once PyPi is updated.

bnmnetp commented:
Thanks for the PR @eoyoa

Just a cursory glance so far, I'll look more carefully in the next couple days.  This is nice as 3.13 has been on my list for a while, but never gets far enough up the stack.

And... you very wisely left the web2py server alone.  Good job.

eoyoa commented:
> I notice that you didn't update the base docker image, so the images are still using 3.11

@bnmnetp, I just committed a change that upgrades the affected projects. Again, I didn't change the ones that didn't need changing (like `web2py`).

However, you still can't test build this PR due to the aforementioned issue (that `rsmanage` needs this PR to be published before it can use `runestone@7.7.4`). I'm thinking it would be nice for a future QOL change to make `poetry build-project` include local versions of `runestone` under a certain `dev` flag or whatever, but it seems:

1. Annoyingly hard, and
2. Potentially out of scope for this PR.

Would like your input!

bnmnetp commented:
Thanks.

I think if you do `poetry env use /path/to/python3.11` from the main rs folder you can then interactively test a bunch of things using `dstart` to at least start a bunch of services.

alternatively it would not be hard for me to do a new release of the interactives so we can move forward.

bnmnetp commented:
Alright, this might be a bit messy for you but hopefully a rebase fix it.

I was going to release 7.7.4 today anyway with a bunch of other bug fixes.  So I went ahead and bumped asyncpg as well.  Which of course cascaded to bumping asyncpg everywhere.

eoyoa commented:
Alright, @bnmnetp, seems to work!

For a summary of updates:

- `Sphinx` is upgraded to be at least version 6.2.0: see https://github.com/sphinx-doc/sphinx/issues/10440#issuecomment-1556180835
- `pydal` is upgraded to be at least version `20241204.1`: see https://github.com/web2py/pydal/pull/717#issuecomment-2469528208
- `pandas` is upgraded to be at least version 2.2.3: see [the relevant changelog](https://pandas.pydata.org/docs/whatsnew/v2.2.3.html)
- `numpy` is upgraded to be at least version 2.1.0: see [the relevant changelog](https://numpy.org/news/#numpy-210-released)
- `asyncpg` is upgraded to be at least 0.30.0: see https://github.com/MagicStack/asyncpg/pull/1188
- Replaced deprecated libraries with `python-deadlib` equivalents
- Removed unused `asyncore` imports in some Runestone components

If you'd like to test it on your end before merging, feel free; you'd have to do the [previously mentioned workaround](https://github.com/RunestoneInteractive/rs/pull/736#issuecomment-2949845474) in order to get it to work locally though.

bnmnetp commented:
OK I've looked at this a bit more.  We have been on sqlalchemy 1.4.x for a very long time.  pandas >= 2.0 needs sqlalchemy >= 2.0. The move to sqlalchemy 2.0 looks like a lot of work and is outside the scope of this PR.

eoyoa commented:
I see. Should we instead try to target Python 3.12 and just say that Python 3.13 is unsupported as of now?

bnmnetp commented:
My quick experiments are indicating that this might not be as bad as it was when I first looked into it.  I think they have done a lot more work to make the transition easier.    The book server needs more testing but my initial build indicates that things are working.  

I'm sure there are edge cases to find, but this might only be a days worth of work instead of weeks.

bnmnetp commented:
OK, Things are fine with sqlalchemy and pandas now.  Hopefully that will pave the way for you to rebase and we can be on 3.13

eoyoa commented:
Ok, hopefully that fixed things; the rebase was a bit of a headache.

bnmnetp commented:
I merged this locally and pushed to main.  Did not automatically close, so closing manually.
